src/scalar/blockbt_mod.F90

```diff
@@ -0,0 +1,205 @@
+MODULE blockbt_mod
+    USE core_mod
+    USE ib_mod, ONLY: ctof, parent, ib
+
+    IMPLICIT NONE(type, external)
+    PRIVATE
+
+    PUBLIC :: blockbt
+
+CONTAINS
+    SUBROUTINE blockbt(bt_f)
+        ! This routine derives a blocking field BT for the scalar
+        ! from the the blocking field BP for the pressure.
+
+        ! Subroutine arguments
+        TYPE(field_t), INTENT(inout) :: bt_f
+
+        ! Local variables
+        TYPE(field_t) :: hilf_f
+        TYPE(field_t), POINTER :: bp_f
+        REAL(realk), POINTER, CONTIGUOUS :: bt(:, :, :), bp(:, :, :), &
+            hilf(:, :, :)
+        INTEGER(intk) :: i, igrid, ilevel
+        INTEGER(intk) :: kk, jj, ii
+        INTEGER(intk) :: nfro, nbac, nrgt, nlft, nbot, ntop
+
+        CALL get_field(bp_f, "BP")
+        CALL hilf_f%init("HILF")
+
+        ! Checking if any neighboring cell (shared face) has bp=1
+        ! [only those cells can have a flux stencil associated with them]
+        DO i = 1, nmygrids
+            igrid = mygrids(i)
+            CALL get_mgdims(kk, jj, ii, igrid)
+            CALL bt_f%get_ptr(bt, igrid)
+            CALL bp_f%get_ptr(bp, igrid)
+            CALL blockbt_grid(kk, jj, ii, bt, bp)
+        END DO
+
+        DO ilevel = minlevel, maxlevel
+            CALL connect(ilevel, 2, s1=bt_f%arr, corners=.TRUE.)
+        END DO
+
+        ! Opening the cells near the parent boundary where the coarse grid
+        ! allows a scalar flux
+        DO ilevel = minlevel, maxlevel
+            hilf_f%arr = -1.0
+            CALL ctof(ilevel, hilf_f%arr, bt_f%arr)
+
+            DO i = 1, nmygridslvl(ilevel)
+                igrid = mygridslvl(i,ilevel)
+
+                CALL get_mgbasb(nfro, nbac, nrgt, nlft, nbot, ntop, igrid)
+                CALL get_mgdims(kk, jj, ii, igrid)
+                CALL bt_f%get_ptr(bt, igrid)
+                CALL hilf_f%get_ptr(hilf, igrid)
+
+                CALL open_scalarflux_boundary(kk, jj, ii, bt, hilf, &
+                    nfro, nbac, nrgt, nlft, nbot, ntop)
+            END DO
+
+            CALL connect(ilevel, 2, s1=bt_f%arr, corners=.TRUE.)
+        END DO
+
+        CALL hilf_f%finish()
+    END SUBROUTINE blockbt
+
+
+    PURE SUBROUTINE blockbt_grid(kk, jj, ii, bt, bp)
+        ! Subroutine arguments
+        INTEGER(intk), INTENT(in) :: kk, jj, ii
+        REAL(realk), INTENT(out) :: bt(kk, jj, ii)
+        REAL(realk), INTENT(in) :: bp(kk, jj, ii)
+
+        ! Local variables
+        INTEGER(intk) :: i, j, k
+
+        DO i = 2, ii-1
```

Comment:
This implementation has a strong focus on robustness and is supposed to ensure mass conservation under all circumstances. There is surely space for refinement (of course not now...) . But maybe that is worth a comment in the code.

Reply:
Noted. Will add a comment.